doc/design/workflow_pacakge.md

```diff
@@ -0,0 +1,85 @@
+# Design Doc: Workflow Package
+
+## Motivation
+
+SQLFlow translates a SQL program, perhaps with extended SQL syntax for AI, into a workflow. Tekton/Argo are Kubernetes native workflow engine when deploying SQLFlow on Kubernetes, SQLFlow leverages Argo/Tekton to do the workflow management.
+
+SQLFlow supports Argo/Tekton as the workflow backend and maybe more in the future. It's different to communicate with the theses workflow engine, .e.g. Argo and Tekton
+are the different CRD on Kubernetes, they are different YAML spec, so it's necessary to orignize a seperate pacakge `workflow` to communicate the workflows with an uniform interface.
+
+## Design
+
+To implement the above motivation, the `workflow` package should include the following functionalities:
+
+1. `CodeGen`: Go interface to generate the Fluid/Argo program to generate workflow YAML.
+1. `Submit/Fetch`: APIs to submit the workflow and trace the status of the workflow step.
+
+We propose the following code structure:
+
+``` text
+workflow/
+        |-workflow.go           # workflow interface
+        |-argo/                 # submit/trace argo workflow via k8s API
+        |-tekton/               # submit/trace tekton workflow via k8s API
+        `-codegen/
+                |-fluid/        # generate Tekton YAML using Fluid
+                `-couler/       # generate Argo YAML using Couler
+```
+
+### Workflow Codegen
+
+Couler/Fluid lets users write Argo and Tekton workflows in Python rather than YAML. Also, the Python code is easier to read and code review.
+SQLFlow implements Fluid `Codegen` to translate the `[]ir.SQLFlowStmt` into Python code, the interface can be like：
+
+``` golang
+type Codegen interface {
+  GenCode([]ir.SQLFlowStmt) string
+  Compile(string) string
```

Comment:
Hard to say running Python to get YAML is "Compile".

Reply:
Maybe can add one common function like `GenYaml(sqlProgram string) string` to directly get YAML from the SQL program.

Reply:
Thanks @typhoonzero , I have thought about using one common function, I prefer two separated functions just because:
1. Fluid is Python code,  so `FluidCodegen` generates Python code instead of YAML code.
1. two separated stage is also easy to debug.

Reply:
Rename `Compile` to `GenYAML` to be more meaningful.